File: Binary-Sreach-1D/Lowerbound.java
// WAP to find the lower bound in array using binary sreach
import java.util.Scanner;

public class Lowerbound {

    // Function to find the lower bound index using binary search
    public static int lowerBound(int[] arr, int target)
    {
        int low = 0;
        int high = arr.length - 1;
        int ans = arr.length; // Default to arr.length if no suitable element is found

        while (low <= high)
        {
            int mid = (low + high) / 2;
            // If current element is greater than or equal to target, it may be a lower bound
            if (arr[mid] >= target)
            {
                ans = mid;         // Update answer
                high = mid - 1;    // Look for a smaller index on the left
            }
            else
            {
                low = mid + 1;     // Target is greater, search right half
            }
        }
        return ans;
    }

    public static void main(String[] args)
    {
        Scanner sc = new Scanner(System.in);

        // Input array size
        System.out.print("Enter the size of the array: ");
        int n = sc.nextInt();

        // Input array elements (assume sorted)
        int[] arr = new int[n];
        System.out.print("Enter " + n + " sorted elements: ");
        for (int i = 0; i < n; i++)
        {
            arr[i] = sc.nextInt();
        }

        // Input target value to find lower bound
        System.out.print("Enter the target value: ");
        int target = sc.nextInt();

        // Call lowerBound function
        int lbIndex = lowerBound(arr, target);

        // Display the result
        if (lbIndex < n)
        {
            System.out.println("Lower bound of " + target + " is at index: " + lbIndex + " (Element: " + arr[lbIndex] + ")");
        }
        else
        {
            System.out.println("No lower bound found. All elements are smaller than " + target);
        }
    }
}

File: Binary-Sreach-1D/Lowerbound.py
# WAP  Given a sorted array of N integers and an integer x, write a program to find the lower bound of x.
# leetcode: https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/
# geeksforgeeks: https://www.geeksforgeeks.org/find-first-and-last-positions-of-an-element-in-a-sorted-array/
# coding ninja: https://www.codingninjas.com/codestudio/problems/first-and-last-positions-of-an-element-in-a-sorted-array_1082549

def lower_bound(arr, x):
    # Initialize the left and right pointers
    left = 0
    right = len(arr)

    # Perform binary search to find the lower bound
    while left < right:
        mid = (left + right) // 2

        if arr[mid] < x:
            left = mid + 1
        else:
            right = mid

    return left

nums=int(input("Enter the number of elements in the array: "))
arr=[]

for i in range(nums):
    element=int(input(f"Enter element {i+1}: "))
    arr.append(element)

x=int(input("Enter the integer x to find its lower bound: "))
result = lower_bound(arr, x)

print(f"The lower bound of {x} in the array is at index: {result}")